### DISCUSSION TITLE
question: can I construct a GEM from a list of KO identifiers?

### DISCUSSION THREAD

**Discussion by silvtal** [SysBioChalmers/RAVEN#410] [Help me!] [answered]
Thank you for developing the RAVEN toolbox

I have a list of KO identifiers from an already functionally annotated genome (used MicrobeAnnotator). My question is - seeing that this toolbox constructs GEMs from the KEGG databases, would it be possible to directly "feed" it the list of KOs?

**Comment by edkerk** [comment, marked as answer]
Interesting idea! There is no straightforward way to do this, but with a little bit of coding you could write a function that can do this.

To get started, you first want to generate the global KEGG model, containing all KEGG reactions:

`keggModel = getModelFromKEGG();`

Then you want to extract the KEGG Orthology annotations:

`KOannot = extractMiriam(keggModel.rxnMiriams, 'kegg.orthology');`

Let's remove all reactions without a KEGG Orthology annotation (from the keggModel and your KOannot list), as you would not have any matching genes for those reactions anyway:

```
noKO = cellfun(@isempty, KOannot);
keggModel = removeReactions(keggModel, noKO, true, true, true);
KOannot(noKO) = [];
```

Now comes the part where you need to write a custom function. What you basically want to do is go through the KOannot list and replace the KO numbers with the gene identifiers and convert this into gene associations (in `model.grRules` format). How easy this is depends on how unique the KO-gene matches are:

1. Does every gene have one KO and does every KO have only one gene?
2. Can a gene be annotated with multiple KOs?
3. Can a KO ben annotated to multiple genes identifiers?

With some combination of `ismember`, `containers.Map` or maybe some old-fashioned loops you can do this, but it very much depends on the layout of your input data. Do you know whether scenario 1, 2 and/or 3 is true for your data? Feel free to post a short sample of what the table looks like.

In the end, if you've replaced the KOs with the gene identifiers and formatted this as a `model.grRules` field, you can use `standardizeGrRules` to make the `rxnGeneMat` field.

**Reply by silvtal** [reply]
Thank you for your fast and very complete reply! It's a relief to know that it is indeed possible. I'm trying a different method, but if I can't make it work I will follow your detailed instructions. Thank you so much!!
